### ISSUE TITLE
support option passthrough (additionalProperties) in schematics and builders

### ISSUE THREAD

**Issue by cladera** [angular/angular-cli#15565]
# 🚀 Feature request

### Command (mark with an `x`)

```
- [ ] new
- [ ] build
- [ ] serve
- [ ] test
- [ ] e2e
- [x] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [x] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc
```

### Description
I want to create schematics and builders that support options not explicitly defined in the schema.  

For example, I want to create a builder that spawns a command just passing through the options from the cli. Pretty much like the [example](https://angular.io/guide/cli-builder#schedule-and-run) in angular.io website although instead of defining `command` and `args` options I'd like the user to type (or define in a target) the options as she or he were executing the command directly. 

### Describe the solution you'd like
According to json-schema's draft, [it is possible](https://json-schema.org/understanding-json-schema/reference/object.html#pattern-properties) to combine `properties`, `propertiesPatterns` and `additionalProperties` to define a schema that allows any property and even control the allowed types.

Currently, the cli only considers `boolean` as the possible type of the property `additionalProperties`. 

https://github.com/angular/angular-cli/blob/fe8666073b4f357d4614bf90ac39225980ebe24c/packages/angular/cli/models/architect-command.ts#L244

I'd suggest to make changes in https://github.com/angular/angular-cli/blob/fe8666073b4f357d4614bf90ac39225980ebe24c/packages/angular/cli/utilities/json-schema.ts#L136 to consider an object as possible value of the property `additionalProperties` and parse the options not included in the `properties` object using its value as definition. 

So, if I create a builder with a schema as follows:

```
 {
  "$schema": "http://json-schema.org/schema",
  "title": "My builder",
  "description": "Executes a command",
  "type": "object",
  "additionalProperties": {
    "type": "string"
  },
  "properties": {
    "command": {
      "type": "string",
      "description": "Command to be executed"
    }
  },
  "required": [
    "command"
  ],
  "definitions": {
  }
}
```

And I pass the following options to the method `parseJsonSchemaToOptions`:

```
['--command', 'my-script', '--foo', 'bar']
```

The method returns the following array:

```
[
 {
   "name": "command",
   "type": "string",
   "required": true,
   "hidden": false
 }
]
``` 

Ignoring the option `--foo`. 

In my option when additionalProperties is an object it should use its value to define the rest of properties returning the following array:

```
[
 {
   "name": "command",
   "type": "string",
   "required": true,
   "hidden": false
 },
 {
   "name": "foo",
   "type": "string",
   "required": false,
   "hidden": false
]
``` 

### Describe alternatives you've considered
If the property `additionalProperties` is set `true` the options not defined explicitly are "stashed" in a property with the name "--" and therefore builders can access to them. 

However, this is not an intuitive workaround and requires the developers to parse these options (split them when using `=` or pair them when using them separately).

**Comment by flash-me** [comment]
This is pretty much what I have been asking in #15547

Nonetheless, I like it...

**Comment by clydin** [comment]
If the goal is to spawn a command, wouldn't the raw arguments array be preferred?  This could then be passed directly to the spawn/exec call.  In an object form this would be far more complicated.

This would also be complicated by the presence of anonymous options; both by themselves and in combination with flag options. For example, how should this be interpreted: `--foo bar baz`?  Is `bar` actually a value for `foo` or is `foo` a boolean flag.  And that also leaves `baz` which would need to be put in the `--` special array property.  Even with the proposed functionality, that property would always need to be checked.

**Comment by cladera** [comment]
@clydin Yeah, it would be better indeed for the developer who implements the builder but I don't think it is better for the user of that builder. In my case, I would expect a command to accept arguments not an argument that is an array of arguments. 

Besides, in some cases I would want some arguments to be required. For example right now I'm implementing a builder to deploy serverless services. So take a look to how a serverless deploy command looks like:

```
sls deploy --region eu-west-1 --foo bar
```
Here, the argument `--region` is a built-in serverless cli argument whereas `--foo` is a custom option defined by the developer in the `serverless.yml` file.

Imagine I need `--region` to be defined always and restrict its possible values (list of aws regions). If I choose to receive all the arguments in an `--args` option I will need to find the region option in the list, parse its value and throw an exception in case is not correct, which is exactly what I get if I define the option in the schema and let the schema validator do the job for me.

Of course I could define  `region` in the schema and have another property for the custom options, but then I'll have to combine two sets of options, the options parsed by the schema validator and the options received in the custom options argument. Not a great deal, but still this is my example, I'm sure there are others far more complex.

Anyway, answering your questions:

* I think we can asume arguments begin with either '--' or '-', which as you well say `baz` cannot be parsed therefore it should go to the `--` array. The builder will decide what to do with everything that went to the `--` array.
* One of the great things of `additionalProperties` feature is it can define types. We can force a certain type or use operators such as `oneOf`. If `--foo` cannot be a boolean, we can determine for sure `bar` is its value. If `--foo` can be a boolean then I think the algorithm should prioritize leaving the least number of arguments in the `--` special array. For example, if `--foo` is defined to be either a boolean or a string I think we should pick `bar` as the string value of `--foo` instead of discarding `bar` and consider `--foo` a flag. 

I'm pretty sure we could come up with an algorithm that covers all this cases and more. 

There is another discussion here beyond the example I could come up with: In my opinion, if the standard, in this case I mean json-schema, describes how to use `additionalProperties` I think builders and schematics should be loyal to that definition and let developers decide how they want to make use of this feature. As I see it now the implementation of `additionalProperties` is incomplete.

**Comment by clydin** [comment]
A builder schema's intended and primary purpose is to described and validate the actual options object passed to the builder.  The full depth and breadth of JSON schema (including additional property behavior) is available for use in this regard.  As a convenience, a limited set of well-defined properties contained within the schema are also converted (where possible) to command-line option definitions.  Command-line options are basic in behavior and cannot adequately express the full feature set of a JSON object or schema.  This automatic conversion may also be augmented in the future to facilitate an opt-in mechanism with a more concrete definition as an extension to the JSON schema.

It is also important to note the negative effect on usability that allowing additional properties provides.  Misspelled options can result in unexpected and potentially damaging behavior as the system will not error if it does not understand an option.  Take the deploy example above (`--region eu-west-1 --foo bar`),  what if additional properties were allowed and a user typed this instead: `--regon eu-west-1 --foo bar`?  If region happened to have a default this could deploy to the wrong region potentially deploying the wrong code to the wrong region.  This may be benign or it could cause a catastrophic failure to the fielded system.  Errors should ideally be caught as early as possible.

Regardless, with the use of the schema highlighted above (with the additional properties as strings), the current code should allow what is begin attempted.  If the user adds additional properties to the builder options in the workspace file, they will become properties on the option object given to the builder.  If the user adds the options onto the command line, they will be received by the builder in their raw form in the `--` special property.  Upon execution, the builder would take any unknown properties from the option object and convert them into command line argument elements.  It would then append the `--` special property to this list of elements and then use it to execute the command.

A common pattern employed to allow custom key/value pairs to be leveraged on the command line is to use a well-defined option to allow passing such values to the system (clang/gcc do this for instance).  As an example, `-Dfoo=bar`/`--define foo=bar` would add `foo` with a value of `bar` to a value map.  This is something that could potentially be supported as it would be useful in other areas as well.

**Comment by cladera** [comment]
@clydin Thank you for the clarification. I totally see your point. 

I'm already using the approach you point out. I expect all the custom properties to be at the special property `--` and as I said, it is not a big deal.

What it is a big deal is the time I had to spend figuring out why setting `additionalProperties` to an object gave me cli errors. There is no documentation, or at least I couldn't find any, explaining the limitations of the implementation of json-schema. I went to json-schema documentation, find exactly what I though it was best solution for the problem I was solving but I couldn't make it work. That's when I though it could be a nice addition to the project. 

But I agree with you, other major challenges may arise and perhaps it is not worth it, I don't know... Although I still think this should be at least documented or handled with a better error message. If you all think this should be considered a bug I can open a new issue.
